### PR TITLE
Add a new shear benchmark for instantaneous and time-dependent flow

### PR DESCRIPTION
**Pull request by gassmoeller**
This is a shear benchmark in a box we recently included in a manuscript. It is useful for benchmarking instantaneous and time-dependent shear flow. The text and detailed derivation is in the manuscript, which will be available soon.

### CODE REVIEW

File: benchmarks/rigid_shear/instantaneous/rigid_shear.prm [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# A description of the SolKZ benchmark for which a known solution
```
Comment (tjhei) [inline benchmarks/rigid_shear/instantaneous/rigid_shear.prm]: update comment?

File: benchmarks/rigid_shear/plugin/rigid_shear.cc [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                                      "rigid shear forcing term",
+                                      "A particle property that represents the force term in the Stokes equation (density * gravity). "
+                                      "The reason this is implemented as a separate particle property is that the gravity becomes "
+                                      "infinite at x=0 and x=1, while (density * gravity) always remains finite. Therefore, "
+                                      "interpolating (density * gravity) does not break down at the model edges, while interpolating "
+                                      "density and then multiplying with the analytical density does.")
+
+    ASPECT_REGISTER_POSTPROCESSOR(RigidShearPostprocessor,
+                                  "rigid shear",
+                                  " The implementation of error evaluators that correspond to the "
```
Comment (tjhei) [inline benchmarks/rigid_shear/plugin/rigid_shear.cc]: remove leading space

File: benchmarks/rigid_shear/time-dependent/rigid_shear.prm [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# A description of the SolKZ benchmark for which a known solution
```
Comment (tjhei) [inline benchmarks/rigid_shear/time-dependent/rigid_shear.prm]: update please

**Comment by gassmoeller** [comment]
Done. Ready from my side.

File: benchmarks/rigid_shear/time-dependent/run_all_models [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+
+# Run all models in this directory. Note that resolution 8 and ppd 96 requires 192-384 GB of RAM
```
Comment (jdannberg) [inline benchmarks/rigid_shear/time-dependent/run_all_models]: But the biggest number in the list is ppd 80, not 96 and the resolution just goes up to 6? So maybe say instead what the largest numbers are that can be run on a laptop?

File: benchmarks/rigid_shear/time-dependent/run_all_models [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        for integrator in "rk2" "rk4"; do # "rk2" "rk4"
+
+        echo "subsection Discretization" > current.prm
+        echo "  set Stokes velocity polynomial degree = $stokes_degree" >> current.prm
+        echo "  set Use locally conservative discretization = $discontinuous_pressure" >> current.prm
+        echo "end" >> current.prm
+
+        number_of_particles=`echo "$particles_per_direction * $particles_per_direction * 2 ^ (2 * $refinement)" | bc -l`
+        #ppc=`expr $particles_per_direction \* $particles_per_direction`
+        #min_ppc=`expr $ppc / 2`
```
Comment (jdannberg) [inline benchmarks/rigid_shear/time-dependent/run_all_models]: Do you need these lines?

File: benchmarks/rigid_shear/time-dependent/run_all_models [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        echo "subsection Postprocess" >> current.prm
+        echo "  subsection Particles" >> current.prm
+        echo "  set Number of particles = $number_of_particles" >> current.prm
+        echo "  set Particle generator name = $generator" >> current.prm
+        #echo "  set Minimum particles per cell = $min_ppc" >> current.prm
+        #echo "  set Minimum particles per cell = $min_ppc" >> current.prm
+        #echo "  set Load balancing strategy = add particles" >> current.prm
```
Comment (jdannberg) [inline benchmarks/rigid_shear/time-dependent/run_all_models]: Is this useful to have?

**Comment by gassmoeller** [comment]
I addressed all comments and added a README file. Ready for another review.

**Comment by gassmoeller** [comment]
Rebased and Indented after the changed to the indent target. Someone please have another look.

File: benchmarks/rigid_shear/instantaneous/creates_formatted_tables.sh [hunk 7]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Comment (bangerth) [inline benchmarks/rigid_shear/instantaneous/creates_formatted_tables.sh]: I think I would call this file `create_formatted_tables.sh` instead.

File: benchmarks/rigid_shear/plugin/rigid_shear.cc [hunk 8]
```diff
@@ -0,0 +1,362 @@
+#ifndef ASPECT_RIGID_SHEAR_H
```
Comment (bangerth) [inline benchmarks/rigid_shear/plugin/rigid_shear.cc]: Do we add copyright headers to the plugin .cc files?

File: benchmarks/rigid_shear/time-dependent/creates_formatted_tables.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
```
Comment (bangerth) [inline benchmarks/rigid_shear/time-dependent/creates_formatted_tables.sh]: same naming issue here

File: benchmarks/rigid_shear/time-dependent/creates_formatted_tables.sh [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+grid_resolution=('3' '4' '5' '6')
+interpolators=('cell_average' 'bilinear_least_squares')
+integrators=('rk2' 'rk4')
+stokes_degrees=('2' '3')
+
+for stokes_degree in ${stokes_degrees[@]}; do
+  for interpolator in ${interpolators[@]}; do
+  for integrator in ${integrators[@]}; do
```
Comment (bangerth) [inline benchmarks/rigid_shear/time-dependent/creates_formatted_tables.sh]: If you indent line 9 by one character less, then things still look nicely staggered :-) If you don't want to do that, then just indent this and the following lines by two more spaces.

File: benchmarks/rigid_shear/time-dependent/results_Q2_bilinear_least_squares_rk2.dat [hunk 11]
```diff
@@ -0,0 +1,28 @@
+.12500000000000000000 16 7.380191e-04 1.882807e-02 6.335796e-04 2.614558e-02 4.380316e-03
```
Comment (bangerth) [inline benchmarks/rigid_shear/time-dependent/results_Q2_bilinear_least_squares_rk2.dat]: same header as in the time independent case?

Comment (bangerth) [reply benchmarks/rigid_shear/time-dependent/results_Q2_bilinear_least_squares_rk2.dat]: same for the other files here

File: benchmarks/rigid_shear/time-dependent/results_Q2_cell_average_rk4.dat [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+.06250000000000000000 100 5.135545e-03 1.623873e-02 4.482334e-03 2.099084e-02 4.003650e-02
+.03125000000000000000 100 1.310482e-03 4.108976e-03 1.143845e-03 5.341148e-03 2.004119e-02
+.01562500000000000000 100 3.260729e-04 1.019699e-03 2.853234e-04 1.334361e-03 1.002419e-02
+.12500000000000000000 225 2.080125e-02 6.589018e-02 1.821708e-02 8.479786e-02 7.971176e-02
+.06250000000000000000 225 5.103925e-03 1.612663e-02 4.453907e-03 2.079700e-02 4.002729e-02
+.03125000000000000000 225 1.274968e-03 4.027111e-03 1.110896e-03 5.191946e-03 2.003545e-02
+.01562500000000000000 225 3.303261e-04 1.032383e-03 2.882911e-04 1.343536e-03 1.002037e-02
+.12500000000000000000 400 2.040231e-02 6.486823e-02 1.784575e-02 8.321215e-02 7.970116e-02
+.06250000000000000000 400 5.257560e-03 1.655133e-02 4.591423e-03 2.137466e-02 4.002341e-02
+.03125000000000000000 400 1.304140e-03 4.084387e-03 1.139362e-03 5.308759e-03 2.003366e-02
```
Comment (bangerth) [inline benchmarks/rigid_shear/time-dependent/results_Q2_cell_average_rk4.dat]: I don't think this is quite correct. It should probably be the following:
```suggestion
.03125000000000000000 400 1.304140e-03 4.084387e-03 1.139362e-03 5.308762e-03 2.003366e-02
```

File: benchmarks/rigid_shear/time-dependent/run_all_models [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+# Run all models in this directory. Note that the highest resolutions and ppd require
+# a high-end laptop or better to run (32 GB RAM).
+
+for stokes_degree in 2 3; do #2 3
+  for discontinuous_pressure in false; do # true
+    for refinement in 3 4 5 6; do
+      for particles_per_direction in 4 5 6 7 10 15 20 32 45 64 80; do # 4 5 6 7 10 15 20
+        for generator in "reference cell"; do
+        for interpolator in "cell average" "bilinear least squares"; do # cell average
+        for integrator in "rk2" "rk4"; do # "rk2" "rk4"
```
Comment (bangerth) [inline benchmarks/rigid_shear/time-dependent/run_all_models]: same idea here :-)

**Comment by gassmoeller** [comment]
Addressed all the comments.

**Comment by gassmoeller** [comment]
@bangerth: That is correct the description and discussion is in the manuscript that is why I felt duplicating it here is not that useful. Are you ok with the merge?

File: benchmarks/rigid_shear/time-dependent/run_all_models [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        echo "end" >> current.prm
+
+        number_of_particles=`echo "$particles_per_direction * $particles_per_direction * 2 ^ (2 * $refinement)" | bc -l`
+        echo "subsection Postprocess" >> current.prm
+        echo "  subsection Particles" >> current.prm
+        echo "  set Number of particles = $number_of_particles" >> current.prm
+        echo "  set Particle generator name = $generator" >> current.prm
+
+        # If using longer runtimes it is necessary to add particles to
+        # revent empty cells. Compute an appropriate number.
```
Comment (bangerth) [inline benchmarks/rigid_shear/time-dependent/run_all_models]: prevent

File: benchmarks/rigid_shear/time-dependent/run_all_models (hunk 6 above)
Comment (bangerth) [reply benchmarks/rigid_shear/time-dependent/run_all_models]: I think @jdannberg is asking a valid question :-)

**Comment by gassmoeller** [comment]
I have fixed the typo. I would like to keep the commented lines, because in order to run the benchmark a bit longer (e.g. a full revolution to produce nice figures) they are necessary.

**Comment by bangerth** [comment]
Please squash/merge commits and then merge yourself.

**Pull request merged**